docs/guides/pallas_kernels_performance.md

```diff
@@ -0,0 +1,144 @@
+
+* **If Maintainability is a Top Priority**: Pallas kernels are lower-level and can be harder to debug. If the performance gain is marginal, the maintenance overhead may not be worth it.
+
+* **Without a Clear Bottleneck**: Avoid premature optimization. Use Pallas only after profiling has identified a specific, memory-bound bottleneck.
+
+
+### Pallas Kernels in MaxText
+
+MaxText uses custom kernels written in Pallas to achieve high performance on TPUs. These kernels optimize key operations within the model, boosting training and inference speed.
+
+* **Splash Attention**: MaxText can use a Pallas-based implementation of Splash Attention for training language models. This is a memory-efficient attention mechanism that processes calculations in smaller, tiled blocks to avoid memory bottlenecks and accelerate training, especially for long sequences. You can find its implementation in `MaxText/kernels/splash_attention_kernel.py`.
+
+* **Paged and Ragged Attention**: For inference, MaxText uses Paged and Ragged Attention kernels. These are highly efficient for handling batched requests with varying sequence lengths by managing the KV cache in non-contiguous memory "pages" and avoiding padding. This is crucial for high-throughput serving and is used across most models. The implementation can be found in `MaxText/inference/paged_attention.py` and `MaxText/inference/paged_attention_kernel_v2.py`.
+
+* **MegaBlox**: For Mixture-of-Experts (MoE) models such as **Mixtral, Qwen, and Deepseek**, we utilize MegaBlox kernels. These are optimized for the sparse Grouped matrix multiplication (GMM) in MoE layers, enabling efficient routing of tokens to experts. It also support quantization for further performance gains. You can see the implementation in `MaxText/kernels/megablox/gmm.py`.
+
+> This list will evolve; treat it as guidance rather than a contract.
+
+
+## Getting Started with Pallas Kernels
+
+### Writing a Custom Kernel
+
+Pallas kernels are Python functions decorated with `@pallas.kernel` that use JAX APIs to operate on `Ref` objects, which are references to JAX arrays. Inside a kernel, you have direct control over reading from and writing to these references.
+
+**Example Kernel:**
+
+```python
+import jax
+import jax.numpy as jnp
+from jax.experimental import pallas as pl
+
+@pl.kernel
+def add_vectors_kernel(x_ref, y_ref, o_ref):
+  x = x_ref[:]
+  y = y_ref[:]
+  o_ref[:] = x + y
+
+def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:
+  return pl.pallas_call(
+      add_vectors_kernel,
+      out_shape=jax.ShapeDtypeStruct(shape=x.shape, dtype=x.dtype),
+  )(x, y)
+```
+
+For a deeper dive, see the [Pallas Quickstart guide](https://jax.readthedocs.io/en/latest/pallas/quickstart.html).
+
+### Integration Steps
```

Comment:
Could we add a section to profile the pallas kernel? Usually use jax.named_scope(). I will cc you in the thread for instructions.